File: agent/.opencode/tools/search-actors.ts
import { tool } from "@opencode-ai/plugin"
import { $ } from "bun"

// Thin wrapper around tools/search_actors.py — located via the git worktree so
// there are no hardcoded paths. The Python side handles Bluesky auth and the
// Cloudflare-friendly curl transport.
export default tool({
  description:
    "Search Bluesky accounts by name, handle, or bio text. Returns one line " +
    "per account: `@handle (Name) — did · counts · bio`. Bios are searched " +
    "too, so a topic ('permaculture') finds people who describe themselves " +
    "that way, not just accounts named after it. Use it to find who is " +
    "already on Bluesky in a subject area.",
  args: {
    query: tool.schema.string().describe("A name, handle, or bio term"),
    limit: tool.schema.number().describe("Max accounts to return").default(25),
  },
  async execute(args, context) {
    const script = `${context.worktree}/tools/search_actors.py`
    // Guard limit — the model sometimes omits it and a schema default doesn't
    // always reach here, which would send the literal "undefined" (argparse
    // exit 2).
    const limit = Number.isFinite(args.limit) ? Math.trunc(args.limit as number) : 25
    return await $`python3 ${script} ${args.query} --limit ${String(limit)}`.env({ ...process.env, ACTING_SESSION: context.sessionID }).text()
  },
})

File: tools/search_actors.py
#!/usr/bin/env python3
"""Search Bluesky accounts by name, handle or bio text. Prints one line per
account — see actor_index.format_actor for the format.

The bio is searched too, not just the handle and display name, which is what
makes accounts findable by what they say about themselves ("permaculture",
"formerly @foo on instagram") rather than only by what they're called.

    python3 search_actors.py permaculture
    python3 search_actors.py "time banking" --limit 50
"""
import argparse
import sys

from actor_index import print_actors
from bsky import BskyError, get


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("query", help="a name, handle, or bio term")
    parser.add_argument("--limit", type=int, default=25)
    args = parser.parse_args()

    data = get("app.bsky.actor.searchActors", {"q": args.query, "limit": args.limit})
    print_actors(data.get("actors", []))


if __name__ == "__main__":
    try:
        main()
    except BskyError as e:
        print(f"actor search failed: {e}", file=sys.stderr)
        sys.exit(1)
